### PR TITLE
add shellcheck and fix complaints

### PR DESCRIPTION
**Pull request by tuminoid**
Add shellcheck. Fix all the shellcheck complaints, and make shell code consistent Bash.
    
Another PR in project-infra will activate the check for PRs.

Markdownlint is separated to its own PR #413

### CODE REVIEW

**Review by kashifest — DISMISSED** [review]

File: prepare-efi.sh
Comment (elfosardo) [inline prepare-efi.sh:31]: maybe add disable=SC2086 here?
the PACKAGES list will be considered as a single string, afraid this will break the build

File: prepare-image.sh
Comment (elfosardo) [inline prepare-image.sh:19]: same here, probably better add disable=SC2086

File: prepare-image.sh
Comment (elfosardo) [inline prepare-image.sh:32]: ditto

File: prepare-image.sh
Comment (elfosardo) [inline prepare-image.sh:55]: ditto

**Review by elfosardo — CHANGES_REQUESTED** [review]

File: prepare-image.sh
Comment (tuminoid) [reply prepare-image.sh:19]: IRONIC_PKG_LIST is a file name and its read by `<`, so here quotes are correct.

File: prepare-efi.sh
Comment (tuminoid) [reply prepare-efi.sh:31]: PACKAGES is defined above to be a single word `grub2-efi-x86`, so it will not break. Variable naming is misleading.

File: prepare-image.sh
Comment (tuminoid) [reply prepare-image.sh:55]: I'll fix that.

**Review by dtantsur — CHANGES_REQUESTED** [review]
I absolutely like the markdown verification, but the diff in the scripts seems mostly useless. There is no point to enforce ${variables} over $variables except that in a few cases.

Landing this will also turn our life downstream into a nightmare, because no backports will apply cleanly (if you do backporting, you'll have the same problem too).

File: scripts/configure-httpd-ipa.sh
Comment (tuminoid) [reply scripts/configure-httpd-ipa.sh:21]: `==` and `=` are effecively the same thing, except `==` is Bash-only and `=` is POSIX compatible.

That said, the usage of `=` and `==` is  very mixed in the scripts, and I just went consistently for `=`. I can change the other way around, if we end up taking this PR in.

**Comment by dtantsur** [comment]
As a note, the Google style guide is pretty opinionated and may not reflect how most people write scripts. Nor does it reflect the intention of $ vs ${}, which is: ${} is used when the variable name is followed by characters that are valid in a variable name, e.g. `${foo}_bar` (cannot be `$foo_bar`).

**Comment by tuminoid** [comment]
> > Stay consistent with what you find

There is no consistency like I said already, so nothing to stick by.

**Comment by tuminoid** [comment]
/hold

Holding this so it doesn't get merged by drive-by lgtm. @elfosardo and @dtantsur please settle if you are willing to take this in. I'm happy to make adjustments if required, but as stated before, I'm going for consistency, maintainability and readability, preferably organization wide. If this does not work for you, I can drop this, and you can implement it in a way that works for you.

File: scripts/configure-httpd-ipa.sh
Comment (tuminoid) [reply scripts/configure-httpd-ipa.sh]: Fixed all of the `=` to `==` as requested. Also made all `[ .. ]` to `[[ .. ]]` as `[ ` is POSIX and `==` is Bash-only and their combination is not supported (yes, Bash doesn't explode, but officially its undefined combination).

**Comment by tuminoid** [comment]
@dtantsur @elfosardo Its been a week. Have you guys discussed this?

**Comment by tuminoid** [comment]
> @dtantsur @elfosardo Its been a week. Have you guys discussed this?

It's now been two weeks. Can you guys discuss your mutual stance on this?

The question in the end is rather simple: Do you want to take the PR in or not? We have bunch of shell code that is a bit messy, not linted, not consistent. We have here PR cleaning it up for you, maybe a bit opinionated PR but guess what, most of the languages/ecossytems have rather opinionated best practices/style guides these days (gofmt, prettier, k8s and so on), because its effective. They're easy for anyone to work in, you can automate your editors and your CI.

> I have no visibiility into your internal development practices, and in this repo there are no branches. I'm trying to have all repositories under metal3-io to have consistent coding style, basic quality checks in place etc so they are in the long run easier to maintain, developers have safety nets in place and faster feedback cycle.
> 
> If you feel this change is not worth of merging or causes too much trouble for you, then it is your right to decline it, but I'd hope the opposite, that you possibly take this into your backports, and after that there is no more conflicts. But like I said, I got no idea what your setup looks like and if that is feasible.

This still stands as well.

**Comment by tuminoid** [comment]
Took away all curly brace changes etc. @dtantsur PTAL.

/test-ubuntu-integration-main

File: scripts/coreos-ipa-common.sh
Comment (dtantsur) [inline scripts/coreos-ipa-common.sh:26]: I wonder why the split is required...

File: scripts/runhttpd
Comment (dtantsur) [inline scripts/runhttpd:3]: Cannot we follow the recommendation in https://www.shellcheck.net/wiki/SC1090 instead of disabling it?

File: scripts/runhttpd
Comment (dtantsur) [inline scripts/runhttpd:33]: same

**Comment by dtantsur** [comment]
Much better, thanks! I'm not a fan of some changes, but I understand the consistency argument here. Just one question inline.

File: scripts/coreos-ipa-common.sh
Comment (tuminoid) [reply scripts/coreos-ipa-common.sh:26]: `$(coreos_kernel_params)` is subshell call and exporting it on the same line masks the possible error from it.

File: scripts/configure-ironic.sh
Comment (elfosardo) [inline scripts/configure-ironic.sh:14]: you could use curly brackets to group multiple lines and remove the second disable statement, I'm ok either way

File: scripts/configure-ironic.sh
Comment (elfosardo) [inline scripts/configure-ironic.sh:48]: is the split necessary here?

File: scripts/runhttpd
Comment (tuminoid) [reply scripts/runhttpd:3]: `disable=SC1091` needs to be there, since shellcheck won't follow `source=xxx` as `/bin/xxx` is considered external source by shellcheck. We could use `SC1090` recommendation, if it was pointing to non-absolute path.

File: scripts/runhttpd
Comment (tuminoid) [reply scripts/runhttpd:33]: Same as above.

File: scripts/configure-ironic.sh
Comment (tuminoid) [reply scripts/configure-ironic.sh:14]: True. IMO it is easier to understand this way, while being explicit about the ignores.

File: scripts/configure-ironic.sh
Comment (tuminoid) [reply scripts/configure-ironic.sh:48]: No its not, good catch. Fixed.

**Comment by tuminoid** [comment]
I took some more unnecessary changes off, and I think I addressed all comments.

/test-centos-integration-main-ironic-source
/test-ubuntu-integration-main

**Comment by tuminoid** [comment]
Cleura is busted but one more try for today.

/test-centos-integration-main-ironic-source
/test-ubuntu-integration-main

**Comment by tuminoid** [comment]
/cc @dtantsur 

Good for LGTM now?

**Review by lentzi90 — COMMENTED** [review]
/lgtm

**Comment by tuminoid** [comment]
@dtantsur can you PTAL and remove changes requested so we can merge this? All comments should be addressed now.

**Review by dtantsur — APPROVED** [review]

**Pull request merged**
